File: dsa-05-main/cpp/day05/singly_list.cpp
#include <iostream>
using namespace std;

// forward declaration of class
class singly_list;

class node {
private:
	int data;
	node* next;
public:
	node() {
		data = 0;
		next = NULL;
	}
	node(int val) {
		data = val;
		next = NULL;
	}
	friend class singly_list;	
};

class singly_list {
private:
	node* head;
public:
	singly_list() {
		head = NULL;
	}
	void add_last(int val) {
		node* newnode, * trav;
		// alloc new node and init it
		newnode = new node(val);
		// special case: if list is empty, newnode itself is first node
		if (head == NULL)
			head = newnode;
		else {
			// traverse till last node
			trav = head;
			while (trav->next != NULL)
				trav = trav->next;
			// add new node to the next of last node
			trav->next = newnode;
		}
	}
	void display() {
		cout << "LIST: ";
		node* trav = head;
		while (trav != NULL) {
			cout << trav->data << ", ";
			trav = trav->next;
		}
		cout << endl;
	}
};

int main() {
	singly_list l1;
	l1.add_last(10);
	l1.add_last(20);
	l1.add_last(30);
	l1.add_last(40);
	l1.display();
	return 0;
}

File: dsa-05-main/python/day05/singlylist.py
class Node:
	def __init__(self, val=0):
		self.data = val
		self.next = None


class SinglyList:

	def __init__(self):
		self.head = None


	# time complexity - O(n)
	def addLast(self, val):
		newNode = Node(val)
		# special case: if list is empty
		if self.head is None:
			self.head = newNode # newNode itself is first node of list
		else: # general case: if list already have few nodes
			# traverse till last node
			trav = self.head
			while trav.next is not None:
				trav = trav.next
			# add address of newNode into last node's next
			trav.next = newNode


	# time complexity - O(n)
	def display(self):
		print("List: ", end='')
		trav = self.head
		while trav is not None:
			print(trav.data, end=", ")
			trav = trav.next
		print()


def main():
	l1 = SinglyList()
	l1.addLast(11)
	l1.addLast(22)
	l1.addLast(33)
	l1.addLast(44)
	l1.display()


if __name__ == "__main__":
	main()

